up-streamer/src/ustreamer.rs

```diff
@@ -145,37 +152,68 @@ impl ForwardingListeners {
         out_authority: &str,
         forwarding_id: &str,
         out_sender: Sender<Arc<UMessage>>,
+        subscription_cache: Arc<Mutex<SubscriptionCache>>,
     ) -> Option<Arc<ForwardingListener>> {
         let in_comparable_transport = ComparableTransport::new(in_transport.clone());
-
         let mut forwarding_listeners = self.listeners.lock().await;
 
-        let (active, forwarding_listener) = forwarding_listeners
-            .entry((in_comparable_transport.clone(), out_authority.to_string()))
-            .or_insert_with(|| {
-                let forwarding_listener = Arc::new(ForwardingListener::new(forwarding_id, out_sender));
-
-                let reg_res = task::block_on(in_transport
-                    .register_listener(&any_uuri(), Some(&uauthority_to_uuri(out_authority)), forwarding_listener.clone()));
-
-                if let Err(err) = reg_res {
-                    warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} unable to register listener, error: {err}");
-                } else {
-                    debug!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} able to register listener");
-                }
+        if let Some((active, forwarding_listener)) = forwarding_listeners
+            .get_mut(&(in_comparable_transport.clone(), out_authority.to_string()))
+        {
+            *active += 1;
+            if *active > 1 {
+                return None;
+            } else {
+                return Some(forwarding_listener.clone());
+            }
+        }
 
-                (
-                    0,
-                    forwarding_listener,
-                )
-            });
-        *active += 1;
+        let forwarding_listener =
+            Arc::new(ForwardingListener::new(forwarding_id, out_sender.clone()));
 
-        if *active > 1 {
-            None
+        // Perform async registration and fetching
+        if let Err(err) = in_transport
+            .register_listener(
+                &any_uuri(),
+                Some(&uauthority_to_uuri(out_authority)),
+                forwarding_listener.clone(),
+            )
+            .await
+        {
+            warn!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} unable to register request listener, error: {err}");
         } else {
-            Some(forwarding_listener.clone())
+            debug!("{FORWARDING_LISTENERS_TAG}:{FORWARDING_LISTENERS_FN_INSERT_TAG} able to register request listener");
+        }
+
+        match subscription_cache
+            .lock()
+            .await
+            .fetch_cache_entry(out_authority.into())
+            .await
+        {
+            Some(subscribers) => {
+                for subscriber in subscribers {
+                    if let Err(err) = in_transport
+                        .register_listener(&subscriber.topic, None, forwarding_listener.clone())
```

Comment:
If we fail to register the listener IMHO we should also roll-back any book-keeping we have done above and unregister.

Might be a good idea to update the function's signature to be returning `Result<Option<Arc<ForwardingListener>, ForwardingListenerError>`

Where you can then clarify why it failed -- e.g. 

```rust
pub enum ForwardListenerError {
  FailToRegisterNotificationRequestResponseListener
  FailToRegisterPublishListener(UUri) // we can embed the subscriber.topic here which failed
}
```

note that you'll have to impl `Error` for `ForwardListenerError` for this to work.

Reply:
If this looks tricky, we can chat about it.

Reply:
Looks good to me